packages/remix-react/single-fetch.tsx

```diff
@@ -226,22 +225,29 @@ export function addRevalidationParam(
   let genRouteIds = (arr: string[]) =>
     arr.filter((id) => manifest.routes[id].hasLoader).join(",");
 
-  // By default, we don't include this param and run all matched loaders on the
-  // server.  If _any_ of our matches include a `shouldRevalidate` function _and_
-  // we've determined that the routes we need to load and the matches are
-  // different, then we send the header since they've opted-into fine-grained
-  // caching.  We look at the `routeModules` here instead of the matches since
-  // HDR adds a wrapper for `shouldRevalidate` even if the route didn't have one
+  // Look at the `routeModules` for `shouldRevalidate` here instead of the manifest
+  // since HDR adds a wrapper for `shouldRevalidate` even if the route didn't have one
   // initially.
   // TODO: We probably can get rid of that wrapper once we're strictly on on
   // single-fetch in v3 and just leverage a needsRevalidation data structure here
   // to determine what to fetch
-  if (matchedRoutes.some((r) => routeModules[r.id]?.shouldRevalidate)) {
-    let matchedIds = genRouteIds(matchedRoutes.map((r) => r.id));
-    let loadIds = genRouteIds(loadRoutes.map((r) => r.id));
-    if (matchedIds !== loadIds) {
-      url.searchParams.set("_routes", loadIds);
-    }
+  let needsParam = matchedRoutes.some(
+    (r) =>
+      routeModules[r.id]?.shouldRevalidate ||
+      manifest.routes[r.id]?.hasClientLoader
+  );
+  if (!needsParam) {
+    return url;
```

Comment:
We don't bother adding a param if there are no `shouldRevalidate`'s and no `clientLoader`'s